### PR TITLE
Add edit and view changes features

### PR DESCRIPTION
**Pull request by mikeebowen**
## This pull request adds 2 new features: 

- User can edit xml parts in VS Code and those changes will be reflected in the OOXML document. E.G. if a user edits a .docx file in VS Code the changes will be reflected when the document is opened in Word or another document editor.
- After editing from an outside program, xml parts that have been changed will be reflected with a yellow asterisk for edited, a green asterisk for created, and a red asterisk for deleted. If a file has been changed the user can right click for the context menu and click "Compare with Previous" to view a diff of the previous version of the xml part and the current version.
- Tests have been added for the new features.
- Documentation in the README has been updated to show users how to user the new features.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      {
+        arrays: 'always-multiline',
+        objects: 'always-multiline',
+        imports: 'always-multiline',
+        exports: 'always-multiline',
+        functions: 'always-multiline',
+      },
+    ],
+    'func-call-spacing': ['error', 'never'],
+    indent: [2, 2, { SwitchCase: 1 }],
```
Comment (yuenm18) [inline .eslintrc.js:20]:
```suggestion
    indent: ['error', 2, { SwitchCase: 1 }],
```

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        code: 140,
+        tabWidth: 2,
+        ignoreComments: true,
+        ignoreUrls: true,
+        // ignoreStrings: true,
+      },
+    ],
+    'no-console': ['warn', { allow: ['warn', 'error'] }],
+    'no-unused-vars': 'off',
+    quotes: [2, 'single', { avoidEscape: true }],
```
Comment (yuenm18) [inline .eslintrc.js:33]:
```suggestion
    quotes: ['error', 'single', { avoidEscape: true }],
```

File: src/ooxml-viewer.ts [hunk 3]
```diff
@@ -1,99 +1,600 @@
...
+        {
+          location: ProgressLocation.Notification,
+          title: 'OOXML Viewer',
+        },
+        async progress => {
+          progress.report({ message: 'Formatting XML' });
+          const folderPath = join(this.fileCachePath, dirname(fileNode.fullPath));
+          const filePath: string = join(folderPath, fileNode.fileName);
+          await this._createFile(fileNode.fullPath, fileNode.fileName, true);
+          const uri: Uri = Uri.parse(`file:///${filePath}`);
```
Comment (yuenm18) [inline src/ooxml-viewer.ts:110]:
```suggestion
          const uri: Uri = Uri.file(filePath);
```

File: src/ooxml-viewer.ts [hunk 4]
```diff
@@ -1,99 +1,600 @@
...
+  async getDiff(file: FileNode): Promise<void> {
+    try {
+      // get the full path for the primary file and the compare files
+      const filePath = join(this.fileCachePath, dirname(file.fullPath), file.fileName);
+      const compareFilePath = join(this.fileCachePath, dirname(file.fullPath), `compare.${file.fileName}`);
+      // create URIs and title
+      const enc = new TextEncoder();
+      const rightUri = Uri.file(filePath);
+      const leftUri = Uri.file(compareFilePath);
+      const rightXml = await (await workspace.fs.readFile(rightUri)).toString();
```
Comment (yuenm18) [inline src/ooxml-viewer.ts:146]:
```suggestion
      const rightXml = (await workspace.fs.readFile(rightUri)).toString();
```

File: src/ooxml-viewer.ts [hunk 5]
```diff
@@ -1,99 +1,600 @@
...
+      const leftXml = await (await workspace.fs.readFile(leftUri)).toString();
```
Comment (yuenm18) [inline src/ooxml-viewer.ts:147]:
```suggestion
      const leftXml = (await workspace.fs.readFile(leftUri)).toString();
```

File: src/ooxml-viewer.ts [hunk 6]
```diff
@@ -1,99 +1,600 @@
...
+      const fileNode: FileNode | undefined = fileNodes.pop();
+      const filePath: string | undefined = fileNode ? join(this.fileCachePath, dirname(fileNode.fullPath), fileNode.fileName) : undefined;
+      if (fileNode && filePath && existsSync(filePath)) {
+        await this.viewFile(fileNode);
+      }
+    }
+  }
+  /**
+   * Closes all active editor tabs
+   * @method _closeEditors
```
Comment (yuenm18) [inline src/ooxml-viewer.ts:188]:
```suggestion
   * @method closeEditors
```

File: src/ooxml-viewer.ts [hunk 7]
```diff
@@ -1,99 +1,600 @@
...
+   * @returns Promise
+   */
+  private async _createFile(relativePath: string, fileName: string, formatIt = false): Promise<void> {
+    try {
+      const preFilePath = join(this.fileCachePath, relativePath);
+      const folderPath = join(this.fileCachePath, dirname(relativePath));
+      const filePath: string = join(folderPath, fileName);
+      await workspace.fs.createDirectory(Uri.file(folderPath));
+      const file: JSZipObject | null = this.zip.file(relativePath);
+      const text: string = (await file?.async('text')) ?? (await (await workspace.fs.readFile(Uri.file(preFilePath))).toString());
```
Comment (yuenm18) [inline src/ooxml-viewer.ts:319]:
```suggestion
      const text: string = (await file?.async('text')) ?? (await workspace.fs.readFile(Uri.file(preFilePath))).toString();
```

**Comment by yuenm18** [comment]
Thank you for the PR, it looks like these changes will be very useful

I have a couple of general comments:
- I noticed an issue when you click a folder in the "OOXML Viewer." It opens up the first file in the folder rather than just opening the dropdown
- Lets make the next version 1.0.0 and have the changelog file contain only the original entry for version 0.0.1 and a new entry for version 1.0.0 with all your changes since version 0.0.1 because the extension only has version 0.0.1 published
- Lets get rid of the releases folder, I want to tag the commit that the version was built off of and add the vsix file to corresponding the github release rather than committing the visx files
- Can you add an extra newlines between your methods

**Comment by mikeebowen** [comment]
Thanks for the suggestions, obviously .toString() is not asynch 🤦‍♂️. 
- I will look into the folder displaying a file. 
- I can remove the tags and revert to v1.0.0 
- I need to keep the releases folder for now, because some coworkers are dog fooding the extension to help me find bugs and that is how I give them the releases. I don't have any objection to you deleting it after the pull request is approved before publishing, though. 
- I can add extra new lines.
- I used a cli tool called [auto-changelog](https://github.com/cookpete/auto-changelog) to create the changelog. It works really well, but because I ran it on my fork it lists all the commits as from my fork. After the pull request is approved, you should rerun auto-changelog so the changelog has the correct links to the commits in your repo.
I will try and have an updated pull request for you tomorrow or the next day. I'll keep you posted.

**Comment by mikeebowen** [comment]
OK, the remaining issues have been taken care of
- Stopped calling viewFile when the user clicks on a FileNode that is a folder
- Reverted the versions and it now goes from v0.0.1 to v1.0.0 and the CHANGELOG.md has been updated with auto-changelog
- Deleted the releases folder, I can distribute the vsix other ways
- Added lines between the methods

Hopefully that takes care of everything 🙂

**Pull request merged**